rust/connlib/tunnel/src/client.rs

```diff
@@ -1233,13 +1233,104 @@ mod tests {
     }
 
     #[test]
-    fn remove_resources_works() {}
+    fn add_resources_update_works_to_dns() {
+        let mut client_state = ClientState::for_test();
+        client_state.add_resources(&[cidr_resource("10.0.0.0/24"), dns_resource("baz.com")]);
+        client_state.add_resources(&[cidr_resource_dns_id("11.0.0.0/24")]);
+
+        assert_eq!(
+            HashSet::<&ResourceDescription>::from_iter(client_state.resources().iter()),
+            HashSet::from_iter(
+                [
+                    cidr_resource_dns_id("11.0.0.0/24"),
+                    cidr_resource("10.0.0.0/24")
+                ]
+                .iter()
+            )
+        );
+        assert_eq!(
+            HashSet::<IpNetwork>::from_iter(client_state.routes()),
+            expected_routes(vec![
+                IpNetwork::from_str("10.0.0.0/24").unwrap(),
+                IpNetwork::from_str("11.0.0.0/24").unwrap()
+            ])
+        );
+    }
 
     #[test]
-    fn set_resource_works() {}
+    fn remove_resources_works() {
+        let mut client_state = ClientState::for_test();
+        client_state.add_resources(&[cidr_resource("10.0.0.0/24"), dns_resource("baz.com")]);
+        client_state.remove_resources(&[cidr_id()]);
+
+        assert_eq!(
+            HashSet::<&ResourceDescription>::from_iter(client_state.resources().iter()),
+            HashSet::from_iter([dns_resource("baz.com")].iter())
+        );
+        assert_eq!(
+            HashSet::<IpNetwork>::from_iter(client_state.routes()),
+            expected_routes(vec![])
+        );
+    }
 
     #[test]
-    fn set_resource_works_replaces_old_resource() {}
+    fn set_resource_works() {
+        let mut client_state = ClientState::for_test();
+        client_state.set_resources(&[cidr_resource("10.0.0.0/24"), dns_resource("baz.com")]);
+        assert_eq!(
```

Comment:
Nit: Would prefer newlines here to visually split "arrange", "act" and "assert".